src/graphnet/data/extractors/i3extractor.py

```diff
@@ -21,20 +21,35 @@ def __init__(self, name):
         self._name = name
 
     def set_files(self, i3_file, gcd_file):
-        # @TODO: Is it necessary to set the `i3_file`? It is only used in one
-        #        place in `I3TruthExtractor`, and there only in a way that might
-        #        be solved another way.
         self._i3_file = i3_file
         self._gcd_file = gcd_file
         self._load_gcd_data()
 
     def _load_gcd_data(self):
         """Loads the geospatial information contained in the gcd-file."""
-        gcd_file = dataio.I3File(self._gcd_file)
-        g_frame = gcd_file.pop_frame(icetray.I3Frame.Geometry)
-        c_frame = gcd_file.pop_frame(icetray.I3Frame.Calibration)
-        self._gcd_dict = g_frame["I3Geometry"].omgeo
-        self._calibration = c_frame["I3Calibration"]
+        if self._gcd_file is None:
+            i3_file = dataio.I3File(self._i3_file)
+            try:
+                g_frame = i3_file.pop_frame(icetray.I3Frame.Geometry)
+                self._gcd_dict = g_frame["I3Geometry"].omgeo
+            except RuntimeError:
+                self.logger.error(
+                    "No GCD file was provided and no G frame was found, exiting."
+                )
+                raise
+
+            try:
+                c_frame = i3_file.pop_frame(icetray.I3Frame.Calibration)
+                self._calibration = c_frame["I3Calibration"]
+            except RuntimeError:
+                pass
+
+        else:
+            gcd_file = dataio.I3File(self._gcd_file)
+            g_frame = gcd_file.pop_frame(icetray.I3Frame.Geometry)
+            c_frame = gcd_file.pop_frame(icetray.I3Frame.Calibration)
+            self._gcd_dict = g_frame["I3Geometry"].omgeo
+            self._calibration = c_frame["I3Calibration"]
```

Comment:
I have tried to clean this up a little bit to avoid code duplication in defintion of `{c,g}_frame`, `self._calibration`, and `self._gcd_dict`.
```suggestion
        # If no GCD file is provided, search the I3 file for frames containing
        # geometry (G) and calibration (C) information.
        gcd_file = dataio.I3File(self._gcd_file or self._i3_file)
        
        try:
            g_frame = gcd_file.pop_frame(icetray.I3Frame.Geometry)
        except RuntimeError:
            self.logger.error(
                "No GCD file was provided and no G-frame was found. Exiting."
            )
            raise
        else:
            self._gcd_dict = g_frame["I3Geometry"].omgeo
            
        try:
            c_frame = gcd_file.pop_frame(icetray.I3Frame.Calibration)
        except RuntimeError:
            self.logger.warning(
                "No GCD file was provided and no C-frame was found."
            )
        else:
            self._calibration = c_frame["I3Calibration"]
```